File: 백준/Gold/14719. 빗물/빗물.py
import sys


def input():
    return sys.stdin.readline().rstrip()


def solution():
    H, W = map(int, input().split())
    blocks = list(map(int, input().split()))

    left_max = [0] * W
    right_max = [0] * W
    water = 0

    max_height_left = blocks[0]
    max_height_right = blocks[W - 1]

    for i in range(W):
        max_height_left = max(max_height_left, blocks[i])
        left_max[i] = max_height_left

    for i in range(W - 1, -1, -1):
        max_height_right = max(max_height_right, blocks[i])
        right_max[i] = max_height_right

    for i in range(W):
        water += min(left_max[i], right_max[i]) - blocks[i]

    return water


print(solution())

File: 백준/Gold/14719. 빗물/빗물.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const readline = require("readline");
let line = 0;
const lines = [];
const input = () => lines[line++];

function readFile(filePath) {
  const readStream = fs.createReadStream(filePath);
  const rl = readline.createInterface({
    input: readStream,
  });

  rl.on("line", (line) => {
    lines.push(line);
  }).on("close", () => {
    console.log(solution());
    process.exit();
  });
}

readFile(filePath);

function solution() {
  const [H, W] = input().split(" ").map(Number);
  const blocks = input().split(" ").map(Number);
  const stack = [];
  let totalWater = 0;

  for (let i = 0; i < W; i++) {
    const currentHeight = blocks[i];
    let lastFilledHeight = 0;

    while (stack.length) {
      const [lastIndex, lastHeight] = stack.at(-1);
      const waterWidth = i - lastIndex - 1;
      const waterHeight =
        Math.min(currentHeight, lastHeight) - lastFilledHeight;
      const amount = waterWidth * waterHeight;
      totalWater += amount;
      lastFilledHeight += waterHeight;

      if (lastHeight <= currentHeight) stack.pop();
      if (lastHeight >= currentHeight) break;
    }

    stack.push([i, currentHeight]);
  }

  return totalWater;
}
